lib/plugin-base.js

```diff
@@ -0,0 +1,337 @@
+var g = require('./globalize');
+
+module.exports = PluginBase;
+
+function PluginBase(options, name, artifact) {
+  this.options = options || {};
+  this.name = name || options.name;
+  this.artifact = artifact || options.artifact;
+}
+
+PluginBase.prototype.getRootDir = function() {
+  return this.options.rootDir;
+};
+
+PluginBase.prototype.load = function(context) {
+  var rootDir = this.getRootDir() || this.options.rootDir;
+  var env = this.options.env;
+  assert(this.name, 'Plugin name must to be set');
+  debug('Root dir: %s, env: %s, artifact: %s', rootDir, env, this.artifact);
+  var config = {};
+  if (this.options[this.name]) {
+    // First check if options have the corresponding config object
+    debug('Artifact: %s is using provided config obj instead' +
+    ' of config file');
+    config = this.options[this.name];
+  } else {
+    if (this.artifact) {
+      config = this.loadNamed(rootDir, env, this.artifact);
+    }
+  }
+  // Register as context.configurations.<plugin-name>
+  return this.configure(context, config);
+};
+
+PluginBase.prototype.configure = function(context, config) {
+  config = config || {};
+  // Register as context.configurations.<plugin-name>
+  if (!context.configurations) {
+    context.configurations = {};
+  }
+  context.configurations[this.name] = config;
+  return config;
+};
+
+PluginBase.prototype.merge = function(target, config, keyPrefix) {
+  return this.mergeObjects(target, config, keyPrefix);
+};
+
+/**
+ * Load named configuration.
+ * @param {String} rootDir Directory where to look for files.
+ * @param {String} env Environment, usually `process.env.NODE_ENV`
+ * @param {String} name
+ * @returns {Object}
+ */
+PluginBase.prototype.loadNamed = function(rootDir, env, name) {
+  var files = this.findConfigFiles(rootDir, env, name);
+  debug('Looking in dir %s for %s configs', rootDir, this.name);
+  if (files.length) {
+    debug('found %s %s files: %j', env, name, files);
+    files.forEach(function(f) {
+      debug('  %s', f);
+    });
+  }
+  var configs = this.loadConfigFiles(files);
+  var merged = this.mergeConfigurations(configs);
+
+  debug('merged %s %s configuration %j', env, name, merged);
+
+  return merged;
+};
+
+/**
+ * Search `rootDir` for all files containing configuration for `name`.
+ * @param {String} rootDir Root directory
+ * @param {String} env Environment, usually `process.env.NODE_ENV`
+ * @param {String} name Name
+ * @param {Array.<String>} exts An array of extension names
+ * @returns {Array.<String>} Array of absolute file paths.
+ */
+PluginBase.prototype.findConfigFiles = function(rootDir, env, name, exts) {
+  var master = ifExists(name + '.json');
+  if (!master && (ifExistsWithAnyExt(name + '.local') ||
+    ifExistsWithAnyExt(name + '.' + env))) {
+    g.warn('WARNING: Main config file "%s{{.json}}" is missing', name);
+  }
+  if (!master) return [];
+
+  var candidates = [
+    master,
+    ifExistsWithAnyExt(name + '.common'),
```

Comment:
This looks like a new feature, allowing e.g. `server/datasources.common.js`. This has been discussed in https://github.com/strongloop/loopback-boot/issues/80, different postfixes (`all` and `dynamic`) were proposed.


Let's keep this change outside of this huge refactoring please and implement it in a new pull request where it's easier to discuss details.